apps/concierge_site/lib/plugs/validate_auth_redirect.ex

```diff
@@ -0,0 +1,28 @@
+defmodule ConciergeSite.Plugs.ValidateAuthRedirect do
+  @moduledoc """
+  Validates that custom auth redirects (used for the register page) are only
+  going to the configured auth provider and nowhere else. This prevents an open
+  redirect security vulnerability.
+  """
+  @behaviour Plug
+  require Logger
+  import Plug.Conn
+  alias Plug.Conn
+
+  @impl Plug
+  def init(opts), do: opts
+
+  @impl Plug
+  def call(%Conn{query_params: %{"uri" => uri}} = conn, _opts) do
+    base_uri = Application.get_env(:concierge_site, :keycloak_base_uri)
+
+    if String.starts_with?(uri, base_uri) do
```

Comment:
question: is this going to crash in the `?uri=BAD_URL&uri=GOOD_URL` case? I think that results in `uri` here being a list and not a `String.t()`.

Reply:
🤯 Let me debug that and see…

Reply:
It appears that it actually just takes the last one.

Please double-check my work. In a test I called:

```
get(
  conn,
  "/auth/keycloak?uri=https%3A%2F%2Fmalicious-site1.com%2FTEST-BASE-URI%2Fauth%2Frealms%2FMBTA%2Fprotocol%2Fopenid-connect%2Fregistrations%3Fclient_id%3Dt-alerts%26response_type%3Dcode%26scope%3Dopenid&uri=https%3A%2F%2Fmalicious-site2.com%2FTEST-BASE-URI%2Fauth%2Frealms%2FMBTA%2Fprotocol%2Fopenid-connect%2Fregistrations%3Fclient_id%3Dt-alerts%26response_type%3Dcode%26scope%3Dopenid"
)
```

and when I inspect the value of `uri` in the implementation code it is `"https://malicious-site2.com/TEST-BASE-URI/auth/realms/MBTA/protocol/openid-connect/registrations?client_id=t-alerts&response_type=code&scope=openid"` and the plug doesn't crash.

Reply:
I also tested hitting `http://localhost:4005/auth/keycloak?uri=https://www.qualys.com?uri=https%3A%2F%2Flogin.mbta.com%2Fauth%2Frealms%2FMBTA%2Fprotocol%2Fopenid-connect%2Fregistrations%3Fclient_id%3Dt-alerts%26response_type%3Dcode%26scope%3Dopenid&uri=https://www.qualys.com?uri=https%3A%2F%2Flogin.mbta.com%2Fauth%2Frealms%2FMBTA%2Fprotocol%2Fopenid-connect%2Fregistrations%3Fclient_id%3Dt-alerts%26response_type%3Dcode%26scope%3Dopenid` locally and it gave me a `bad request` response.